### PR TITLE
before embedding check that the destination basis matches the operator basis

### LINKED ISSUE #240: embed with false indexing overwrites sub-basis
This bit of code
```julia
using QuantumOptics
bf = FockBasis(5)
bs = SpinBasis(1//2)
σ = sigmam(bs)

embed(bf⊗bs,1,σ)
```
results in
```julia
SparseOperator(dim=4x4)
  basis: [Spin(1/2) ⊗ Spin(1/2)]
  [2, 1]  =  1.0+0.0im
  [4, 3]  =  1.0+0.0im
```
which clearly overwrites the `FockBasis`. This should throw an error.

### PR DESCRIPTION
**Pull request by atombear**
closes #240

### CODE REVIEW

**Comment by david-pl** [comment]
I changed the syntax as discussed in #240. I'm sure things could be more elegant, but at least everything seems to work properly like this. Let me know if you have any suggestions before I merge this.

**Comment by david-pl** [comment]
I didn't go through your code in detail, so correct me if I'm wrong, but it seems like your function handles the case of a single (composite) operator being embedded in a product basis only. This case is covered by the method I wrote as well. So yes, I think this is a bit redundant.

**Comment by atombear** [comment]
Yes, my method handles the case of one operator being embedded in any subspace of a composite hilbert space. so, for 8 qubits, D=2^8, and a CNOT will be embedded appropriately in any 4d supspace, meaning in any 2 indices. I think your method does not distinguish between an operator acting an different, but equivalent, sub-bases. for example, for a 3x5 operator being embedded in a space that is 3x3x5, your method gives the same output whether the indices are [1, 3] or [2, 3], but those should yield different operators.

**Comment by atombear** [comment]
so i think my position is as follows: the original code `(indices, ops)` tried to match every `op` to an `index`, but had the problem that the matching would not bother checking that the `op` matched the basis corresponding to the `index` in which it was being inserted.

the new method `(indices, ops)` tries to do two things:
- assert that the size of the basis matches, thereby fixing the bug
- try to account for composite spaces by index counting

however, there is an oversight in that it assumes operators in composite spaces correspond to consecutive basis indices. i think it is impossible to correct for this without the index wrangling. Additionally, it assumes the following syntax: `[1,2,3], [op12, op3]` to place a composite operator `op12` in the space `[1,2]`, and a third operator `op3` in space `3`. perhaps the syntax `[[1,2],3], [op12, op3]` would be clearer?

the new `(indices, op)` method generically embeds an operator in a composite hilbert space in an even larger hilbert space, by doing the appropriate index wrangling.

i think it is true that the more general and bug-free `(indices, ops)` method can be written in terms of the new `(indices, op)` method, allowing for non-consecutive indices. my proposal is to do this, and i am glad to take on the task if that makes sense to you. additionally, i am happy with any syntax, though the second i think is clearer, eg: `[[1,3],2], [op13, op2]`.

**Comment by atombear** [comment]
i think i have pushed a test-passing working version of what i described. there are a few more tests i'd like to add, but i think the code is at least correct (finally). a review would be much appreciated at this point.

**Comment by david-pl** [comment]
I appreciate all your work. The changes look good and I agree with the new syntax, which makes a lot more sense especially in the case that I oversaw (embedding composite operators on bases such as e.g. [1,3]). I'm also not sure why the tests are failing. Hopefully it was just a hick-up, so I triggered them by closing/re-opening.

File: src/operators.jl [hunk 1]
```diff
@@ -88,22 +89,128 @@ tensor(op::AbstractOperator) = op
...
 
 
+
+"""
+    check_indices(indices::Array)
+
+Determine whether a collection of indices, written as a list of (integers or lists of integers) is unique.
+This assures that the embedded operators are in non-overlapping subspaces.
+"""
+function check_indices(indices::Array)
```
Comment (david-pl) [inline src/operators.jl]: It would make sense to move this function into the `sortedindices` module

File: src/operators.jl [hunk 2]
```diff
@@ -88,22 +89,128 @@ tensor(op::AbstractOperator) = op
...
     @assert length(basis_r.bases) == N
     @assert length(indices) == length(operators)
+
+    # Embed all single-subspace operators.
+    idxop_sb = [x for x in zip(indices, operators) if typeof(x[1]) <: Int64]
+    indices_sb = [x[1] for x in idxop_sb]
+    ops_sb = [x[2] for x in idxop_sb]
+
+    for (idxsb, opsb) in zip(indices_sb, ops_sb)
+        @assert (opsb.basis_l == basis_l.bases[idxsb]) || throw(bases.IncompatibleBases())
```
Comment (david-pl) [inline src/operators.jl]: The `@assert` statement here (and in the next line) is unnecessary

File: src/operators.jl [hunk 3]
```diff
@@ -88,22 +89,128 @@ tensor(op::AbstractOperator) = op
...
+    embed(basis1[, basis2], indices::Vector, operators::Vector)
+
+Embed operator acting on a joint Hilbert space where missing indices are filled up with identity operators.
+"""
+function embed(basis_l::CompositeBasis, basis_r::CompositeBasis,
+               indices::Vector{Int}, op::T) where T<:AbstractOperator
+    N = length(basis_l.bases)
+    @assert length(basis_r.bases) == N
+
+    @assert reduce(tensor, basis_l.bases[indices]) == op.basis_l || throw(bases.IncompatibleBases())
```
Comment (david-pl) [inline src/operators.jl]: Again, `@assert` is not needed here.

File: src/operators.jl (hunk 1 above)
Comment (atombear) [reply src/operators.jl]: great

File: src/operators.jl (hunk 2 above)
Comment (atombear) [reply src/operators.jl]: done

File: src/operators.jl (hunk 3 above)
Comment (atombear) [reply src/operators.jl]: done.

**Pull request merged**